### DISCUSSION TITLE
How to keep source maps private on the live site ?

### DISCUSSION THREAD

**Discussion by davestewart** [getsentry/support#46] [Support] [answered]
Hi there.

I'm working on a Vite/Vue project and we seem to have set up Sentry properly in both Vite (to upload source mpas) and Vue (to report errors.

However, when we go the deployed URL we can open dev tools and see the full source code, I presume because the source maps are also deployed.

I then added a script to both remove the source maps, and scrub the `sourceMappingURL=` comments, but then we don't see the original source code in Sentry, even though they have been uploaded successfully (we can see the script output in Circle CI and the artefacts in Sentry)

In this image, you can see both versions (top = source maps removed, bottom = source maps deployed):

![image](https://user-images.githubusercontent.com/132681/180515817-2ad5e904-96b1-4c67-9370-225eead5d64f.png)

So my question is, how do we see original source maps in Sentry, but not have them reveal source code on the live site?

Thanks.

**Comment by HazAT** [comment, marked as answer]
You can (should) upload the source maps to Sentry directly with `sentry-cli`
https://docs.sentry.io/platforms/javascript/guides/vue/sourcemaps/

We try to automatically fetch it, if they are publically available but if not, `sentry-cli` is the tool you want to use.
We also provide a Webpack plugin that makes it super simple to upload the source maps.

Edit (so we can mark this as answer):
Also, ensure you have a `release` configured during source map upload and during the initialization of the SDK in `Sentry.init`.
We need the `release` in the event and source map to unminify the stacktrace.

**Reply by HazAT** [reply]
@davestewart thanks for the valuable feedback! 
I hope it's appropriate the mark this as the answer for anyone in the future Googling for an answer

**Reply by davestewart** [reply]
Sure!

How I would word it would be probably along the lines of the example I gave in one of the other comments...

To associate incoming errors with the correct uploaded source maps, Sentry uses a config property called `release`.

The only requirement for this property is that it is unique between builds; approaches include using either the `package.json` version, or the current Git commit hash (the setup for this is a bit more complicated, via a plugin or build step).

Note that you must use the same value in both the **build** and **runtime** configurations:

Build:

```js
// /vite.config.js
const pkg = require('package.json')

// Vite Sentry plugin (note, only release config shown)
const sentry = ViteSentry({
  release: pkg.version,
}),

// export final config
export default {
  plugins: [
    sentry
  ]
}
```

Runtime:

```js
// /src/main.js
import pkg from '../package.json'

// Sentry SDK (note, only release config shown)
Sentry.init({
  release: pkg.version,
})
```

**Comment by davestewart** [comment]
Hey Daniel,

Thanks for the response.

Our workflow is this:

- it's Vue project using Vite
- build on Circle CI
- we build both the files and the source maps 
- Use [Vite Plugin Sentry](https://github.com/ikenfin/vite-plugin-sentry) to upload source maps (this works)
- after the build we attempted to remove the source maps (to keep them private) using a Node script

![image](https://user-images.githubusercontent.com/132681/180802937-47343955-7d0f-44df-913f-be47f77ca21e.png)

As illustrated in the previous image, once we remove the source maps, we can see that Sentry doesn't make the link between what was uploaded and the files on the server (`assets/index.21945` vs `src/app/components/layout/MlTopBar`):

![](https://user-images.githubusercontent.com/132681/180515817-2ad5e904-96b1-4c67-9370-225eead5d64f.png)

Can you help me understand what steps or config I'm missing to make this happen?

**Reply by HazAT** [reply]
You need to ensure you configure a `release` during the upload of the source maps (which I think you did) and also set the same release when you `Sentry.init` in your code as an option.

https://docs.sentry.io/platforms/javascript/sourcemaps/troubleshooting_js/
more specifically https://docs.sentry.io/platforms/javascript/sourcemaps/troubleshooting_js/#verify-a-release-is-configured-in-your-sdk

We know it's a PITA to set up once and get it running, we constantly try to improve it but it's very hard to support every possible constellation in JS seamlessly 😅 

As an explanation: We use the release identifier of the uploaded source maps & event to match them up and do the unminifaction.

**Comment by davestewart** [comment]
Setting `release` may be the missing link!

I wasn't quite sure how to determine the value as we're often releasing to dev or staging branches vs a final production branch with a proper semver number.

Let me look into using something like the commit hash, which we set in the AWS secrets manager in the Circle CI build, and so is available as an environment variable in the build.

Should that work, or am I over-thinking / over-engineering this?

Disclaimer: I'm not a dev ops guy!

**Reply by HazAT** [reply]
Any `string` / identifier (it doesn't have to be semver or anything) works, we at Sentry use the git sha
The important part is it has to match, only then can we apply the source maps.

**Reply by davestewart** [reply]
OK, good news - we got it working!

Thanks to your help @HazAT we used the commit hash as I mentioned and can now see source maps in Sentry but not on the public URL.

I have a little bit of feedback about the docs which might be helpful for you guys, but I'll leave it as a main thread comment so if you want to reply it will be easier to track.

Thanks again though! Much appreciated. 🙏

**Comment by AbhiPrasad** [comment]
Just because I'm curious @davestewart - did you walk through https://docs.sentry.io/platforms/javascript/guides/vue/sourcemaps/troubleshooting_js/? I wonder if there are things we can make more clear there.

**Reply by davestewart** [reply]
Hey Abhi,

I hope you had a nice holiday 🏝 !

If you check my other top-level comment, I described my experience of using the help.

I think it might be useful to sit with a new user and watch them use the help in the same way I just did to see how they fare. There is a LOT of it, and it seems to direct the user to some very advanced things before perhaps finishing the basics.

Had I worked though this document? Probably not as much as I should, and as mentioned below there are a lot of terms which perhaps seem ambiguous or unclear such as "artefacts" or "release" (which I feel is more an identifier, than a traditional "release" where I might refer to something with a semver value).

Let's take the following section which might be relevant in my case:

> **Verify a release is configured in your SDK**
>
> For uploaded source maps to be located and applied, the release needs to be created by the CLI or API (and the correct artifacts uploaded with it), and the name of that newly-created release needs to be specified in your SDK configuration.
>
> To verify this, open up the issue from the Sentry UI and check if the release is configured. If it says "not configured" or "N/A" next to Release on the right hand side of the screen (or if you do not see a release tag in the list of tags), you'll need to go back and [tag your errors](https://docs.sentry.io/platforms/javascript/guides/vue/configuration/releases/). If this is properly set up you'll see "Release: my_example_release".

In order:

- I don't know what an SDK is in this context
- I don't know what "creating a release" in this context means
- I'm not using the CLI directly, I'm using the Vite plugin
- I don't know what the API refers to in this case (my guess is that it has something to do with the back end, in which case, why is it in the Vue section?)
- I don't know what an "artefact" is (I think now it's any file, such as the source maps themselves)
- "the name of that newly-created release needs to be specified in your SDK configuration" - again, no idea what an SDK configuration is (I presume now it just means "the Vue build configuration" but it should just say this)
- "open up the issue from the Sentry UI" – what is the UI? Is this the dashboard? Maybe a screenshot would help here
- "or if you do not see a release tag in the list of tags" – I have no idea what a tag is (but if I look on the page, I can indeed see a section called "Tags" and there is one with the release has in!)
- "you'll need to go back and tag your errors" – I click on the link and see nothing in this page about "tagging errors". However, if I search "tag" on that page I see "Setting the release name tags each event with that release name." which doesn't make it much clearer.
- "If this is properly set up you'll see **Release: my_example_release**" – eeek! This is up there with `dsn: https://examplePublicKey@o0.ingest.sentry.io/0` which I just copy/pasted in as I didn't understand it was supposed to be changed and taken from my Dashboard settings.

Looking again at the [Releases and Health](https://docs.sentry.io/platforms/javascript/guides/vue/configuration/releases/) page which was mentioned in that paragraph, it does seem that there is some useful information, but it's not intuitive for a new user:

> A release is a version of your code that is deployed to an [environment](https://docs.sentry.io/platforms/javascript/guides/vue/configuration/environments/). When you give Sentry information about your releases, you can:

Again, assumes I know what all this stuff is

> Bind the Version

Oh wow... this is really getting a bit meta

However the information inside is really what I need!

I think if the pertinent bits were extracted and moved to the Getting Started, this would be useful. Sentry really needs to find the balance between too little and too much though. It could be as simple as:

> To associate incoming errors with commits and uploads, Sentry uses a config value called `release`. The only requirement for this value is that it is unique between builds. We recommend using either the `package.json` `name` and `version` or the current Git commit `hash`. 
>
> You will need to use the same value in both the build and runtime configurations, so sentry can associate build files (i.e. your source maps) and running code (i.e. errors generated and sent to Sentry).

Right now, I have to look in all these places and ignore all the stuff in between:

```
SENTRY FOR VUE

 +- Getting Started                     <-- nothing really in here
 +- Configuration
 |   +- Basic Options                   <-- again, totally swamped with additional information
 |   +- Integrations
 |   +- Environments
 |   +- Releases & Health               <-- some of the information is in here, but very esoteric
 |   +- Sampling
 |   +- Filtering
 |   +- Shutdown and Draining
 |   +- Sentry Testkit
 |   +- Tree Shaking
 +- Usage
 +- Source Maps
 |   +- Generating Source Maps
 |   +- Uploading Source Maps           <-- we use a plugin for this
 |   +- Validating Files
 |   +- Best Practices
 |   +- Troubleshooting Source Maps     <-- there's a LOT of information here that takes you to lots of other places
 +- Troubleshooting
```

I suppose the real takeaway here is that there's just too much documentation for someone who just wants to get up and running.

Once I'm up and running, by all means, I'm happy to dig deep and find out how it all works.

But if i never get up and running, I'm more likely to just jack it and not bother.

FWIW the [documentation of the Vite plugin](https://github.com/ikenfin/vite-plugin-sentry#share-config-with-sentry-client-library) could be a lot better / focused on a new user too.

FWIW I'm an experienced developer and I'm willing to put the work in to get it all working, but it's a bit gruelling when you can't see the wood for the f***ing trees!

👍

**Reply by AbhiPrasad** [reply]
@davestewart Thank you so much for the feedback! I've forwarded it to relevant folks to look at.

We def need to do better here in general with the getting started documentation, your details really help us a lot here.

The vite plugin is not officially supported by Sentry, it's community contributed, but we can help improve the docs there as well.

**Reply by davestewart** [reply]
No problem!

I wonder if part of the getting started there should be a link to a terms page, or a "Sentry 101" ?

Not all of the terms, but at least the basics (so "environment" and "release", but not "traces sample rate") and maybe a visual diagram of how things fit together.

**Comment by davestewart** [comment]
OK, as mentioned above, I thought it would be helpful to leave a bit of feedback about getting to the point where we have things up and running as we wanted.

@HazAT – you mentioned it's impossible to cover all bases, and I totally appreciate that, but here's a little feedback on my experience of using the docs to try and get things set up:

- there does seem to be a lot of documentation when you first arrive
- there are also a lot of terms; dsn, artefact, dist, release, samples, breadcrumbs, etc
- the "getting started" section feels like it doesn't quite have enough information
- the next steps are **many** and feel like information that should be in "getting started" is probably hidden there
- there's not quite enough information about the various terms to help a new user understand what they are; for example a link to the user's [DSN](https://sentry.io/settings/metalinkapp/projects/frontend/keys/) in Sentry settings would be useful – I didn't even know what a DSN was! For all I know it was something like an XML schema...
- the [basic options](https://docs.sentry.io/platforms/javascript/configuration/options/) page is **long**. Surely the basic options should be much more concise? I just want to get up and running, and it seems that I could have done that with just a very few options (dsn, release & integrations, not tracesSampleRate (is this important here?)).
- I think preventing source maps from being accessible on the public server should be considered a prerequisite, not a nice to have, and as such, the basic setup docs should place more onus on this. There are notes about [public](https://docs.sentry.io/platforms/javascript/sourcemaps/) but nothing about private, and I think this is a mistake.
- the docs (which _are_ informative) are written in a way which _assumes_ knowledge; for example exhaustively listing **all** the configuration options and using language which contains many esoteric terms. I feel especially the getting started section could be trimmed down and solely focus on the 4 or 5 key configuration options to prevent overwhelming the user (I realise this might seem counter to my previous point, but I see keeping source maps private as important)
- the [Vue](https://docs.sentry.io/platforms/javascript/guides/vue/) docs don't actually mention `release`. As it seems like a fairly critical option – especially for source maps  perhaps this could be included in all examples (and perhaps deserves more explanation /disambiguation overall, including how to generate a number as you mentioned above, and as I also implemented).
 
I also recognise that you have a large number of platforms to support and being reasonably consistent across all of them is yet another challenge.

Anyway, I hope that was helpful, and thanks again for getting us over the line 🙏
